app/js/GcsPersistor.js

```diff
@@ -0,0 +1,273 @@
+    }
+
+    const meteredStream = PersistorHelper.getMeteredStream(
+      readStream,
+      'gcs.egress' // egress from us to gcs
+    )
+
+    const writeOptions = {
+      // disabling of resumable uploads is recommended by Google:
+      resumable: false
+    }
+
+    if (sourceMd5) {
+      writeOptions.validation = 'md5'
+      writeOptions.metadata = {
+        md5Hash: PersistorHelper.hexToBase64(sourceMd5)
+      }
+    }
+
+    const uploadStream = storage
+      .bucket(bucketName)
+      .file(key)
+      .createWriteStream(writeOptions)
+
+    await pipeline(meteredStream, uploadStream)
+
+    // if we didn't have an md5 hash, we should compare our computed one with Google's
+    // as we couldn't tell GCS about it beforehand
+    if (hashPromise) {
+      sourceMd5 = await hashPromise
+      // throws on mismatch
+      await PersistorHelper.verifyMd5(GcsPersistor, bucketName, key, sourceMd5)
+    }
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'upload to GCS failed',
+      { bucketName, key },
+      WriteError
+    )
+  }
+}
+
+async function getFileStream(bucketName, key, opts = {}) {
+  if (opts.end) {
+    // S3 (and http range headers) treat 'end' as inclusive, so increase this by 1
+    opts.end++
+  }
+  const stream = storage
+    .bucket(bucketName)
+    .file(key)
+    .createReadStream(opts)
+
+  const meteredStream = PersistorHelper.getMeteredStream(
+    stream,
+    'gcs.ingress' // ingress to us from gcs
+  )
+
+  try {
+    await PersistorHelper.waitForStreamReady(stream)
+    return meteredStream
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'error reading file from GCS',
+      { bucketName, key, opts },
+      ReadError
+    )
+  }
+}
+
+async function getFileSize(bucketName, key) {
+  try {
+    const [metadata] = await storage
+      .bucket(bucketName)
+      .file(key)
+      .getMetadata()
+    return metadata.size
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'error getting size of GCS object',
+      { bucketName, key },
+      ReadError
+    )
+  }
+}
+
+async function getFileMd5Hash(bucketName, key) {
+  try {
+    const [metadata] = await storage
+      .bucket(bucketName)
+      .file(key)
+      .getMetadata()
+    return PersistorHelper.base64ToHex(metadata.md5Hash)
+  } catch (err) {
+    throw PersistorHelper.wrapError(
+      err,
+      'error getting hash of GCS object',
+      { bucketName, key },
+      ReadError
+    )
+  }
+}
+
+async function deleteFile(bucketName, key) {
+  try {
+    await storage
+      .bucket(bucketName)
+      .file(key)
+      .delete()
+  } catch (err) {
+    const error = PersistorHelper.wrapError(
+      err,
+      'error deleting GCS object',
+      { bucketName, key },
+      WriteError
+    )
+    if (!(error instanceof NotFoundError)) {
+      throw error
+    }
+  }
+}
+
+async function deleteDirectory(bucketName, key) {
+  if (!key.match(/^[a-z0-9_-]+/i)) {
```

Comment:
For safety, how about we prevent deleting anything more than the content for a specific file:
```
!key.match(/^[0-9a-f]{24}\/[0-9a-f]{24}/i)
```

Reply:
yes, that is better

Reply:
So, when running the acceptance tests I'm seeing URLs such as
```
acceptance_tests_0.9991905513266544/directory/0.23709030119276875-converted-cache/
```
which don't match.

One thing that occurs is that the persistor is somewhat generic, so enforcing this path style (objectid/objectid) here where we're not doing it in other places is a bit awkward. For example, if I want to plug this in to some other service like docstore, where we want to migrate from S3 to GCS, the logic then gets in the way.

Reply:
Replacing the `id = Math.random()`'s in the acceptance tests with `ObjectId()` (which creates a random objectid) should fix most of those failures.

I think it would be easy enough to make this check configurable if we need to use the code elsewhere.  

I do think it's worth having multiple layers of defenses against accidentally deleting our entire filestore bucket as this would not be easy to recover from!